java-sdk/sdk-checkpoints.asciidoc

```diff
@@ -0,0 +1,88 @@
+= Checkpoints in the Connectors SDK
+
+== Use Cases
+
+=== Incremental Re-crawl
+
+Incremental re-crawl can be supported when a Changes API is available (e.g., Jive, Salesforce, OneDrive). When a Changes API is available, it’s necessary to provide an input parameter to be tracked, such as a date, link, page token, or other. The input parameter is generated (retrieved) while running the first job. During the next job, that parameter will be used to query the Changes API and retrieve new, modified, and deleted objects.
+
+The SDK provides a way to store the input parameters, establishing *checkpoints*, and use them in the subsequent jobs.
+
+[[checkpoint-design]]
+==== Checkpoint Design
+
+Fetcher implementations can emit checkpoint messages by calling:
+
+```
+fetchContext.emitCheckpoint(CHECKPOINT)ID, checkpointMetadata);
+```
+
+After the checkpoint is emitted, the SDK controller will handle this message as follows:
+
+* The checkpoint will not be used in the current job.
+* The checkpoint will be stored in the CrawlDB with the appropriate status.
+
+In the next job, the SDK controller will check the CrawlDB to verify if there are checkpoints stored. If checkpoints are available, only the checkpoints will be sent to the fetchers. Checkpoints will be handled by the fetcher which emitted that checkpoint. If checkpoints are not available, all items in the CrawlDB (documents, fetchInputs, errors, etc.) will be sent to the fetchers.
+
+IMPORTANT:  In order to update a checkpoint, it must be emitted with the checkpoint ID. The checkpoint ID is the only way the SDK controller can identify and update a checkpoint.
+
+===== First Job Flow
+
+image:/assets/images/sdkcheck-1stflow.png[First Job Flow]
+
+. The Jobs API sends a start job request to the SDK controller.
+. The SDK controller queries the SDK CrawlDB to check for items.
+.. It’s the first job, so the SDK CrawlDB is empty. The controller will send the initial FetchInput to the fetcher.
+. During the job, the fetcher receives a FetchInput. The fetcher can then emit candidates and/or checkpoints.
+. When the SDK controller receives a checkpoint message, the checkpoint is stored or updated in CrawlDB. It also will process the candidates sent.
+. The SDK controller will not send the checkpoint to the fetcher in the same job.
```

Comment:
```suggestion
. Since checkpoints are designed to be used for incremental crawling, the SDK controller will only send them back to fetchers in subsequent job runs.
```